urdf/src/model.cpp

```diff
@@ -124,50 +145,35 @@ bool Model::initXml(TiXmlElement * robot_xml)
   return Model::initString(ss.str());
 }
 
-bool Model::initString(const std::string & xml_string)
+bool Model::initString(const std::string & data)
 {
   urdf::ModelInterfaceSharedPtr model;
 
-  // necessary for COLLADA compatibility
-  if (IsColladaData(xml_string)) {
-    fprintf(stderr, "Parsing robot collada xml string is not yet supported.\n");
-    return false;
-    /*
-    ROS_DEBUG("Parsing robot collada xml string");
-
-    static boost::mutex PARSER_PLUGIN_LOCK;
-    static boost::scoped_ptr<pluginlib::ClassLoader<urdf::URDFParser> > PARSER_PLUGIN_LOADER;
-    boost::mutex::scoped_lock _(PARSER_PLUGIN_LOCK);
-
-    try
-    {
-      if (!PARSER_PLUGIN_LOADER)
-        PARSER_PLUGIN_LOADER.reset(new pluginlib::ClassLoader<urdf::URDFParser>("urdf_parser_plugin", "urdf::URDFParser"));
-     const std::vector<std::string> &classes = PARSER_PLUGIN_LOADER->getDeclaredClasses();
-     bool found = false;
-     for (std::size_t i = 0 ; i < classes.size() ; ++i)
-       if (classes[i].find("urdf/ColladaURDFParser") != std::string::npos)
-     {
-       boost::shared_ptr<urdf::URDFParser> instance = PARSER_PLUGIN_LOADER->createInstance(classes[i]);
-       if (instance)
-         model = instance->parse(xml_string);
-       found = true;
-       break;
-     }
-     if (!found)
-       ROS_ERROR_STREAM("No URDF parser plugin found for Collada files. Did you install the corresponding package?");
-   }
-   catch(pluginlib::PluginlibException& ex)
-   {
-     ROS_ERROR_STREAM("Exception while creating planning plugin loader " << ex.what() << ". Will not parse Collada file.");
-   }
+  size_t best_score = std::numeric_limits<size_t>::max();
```

Comment:
If I understand properly, a lower score means a better result.  Just out of curiousity, why did you pick a lower score as better?  I think it is slightly more intuitive if a higher score is better, but maybe there is a good reason not to do that.

Relatedly, it looks like if there are several plugins that report the "best" score, the first one wins.  I'm fine with that as a default policy, but we should document that somewhere.

Do you think there is value in letting the caller of the API either define an exclusive list of plugin names to try, or to provide "hints" as to which they prefer?

Reply:
> If I understand properly, a lower score means a better result. Just out of curiousity, why did you pick a lower score as better? I think it is slightly more intuitive if a higher score is better, but maybe there is a good reason not to do that.

I picked it because I was already thinking of using the position of the first recognized XML tag, and it was more straight forward to implement if lower score is better.

> Relatedly, it looks like if there are several plugins that report the "best" score, the first one wins. I'm fine with that as a default policy, but we should document that somewhere.

Documented behavior on `urdf::Model` class in 3adaaba

> Do you think there is value in letting the caller of the API either define an exclusive list of plugin names to try, or to provide "hints" as to which they prefer?

There might be some value on a development machine where a robot description is ambiguous, or an installed plugin takes a really long time to do `might_handle()`. On a robot itself there's less value because presumably someone has chosen to install only the plugins the robot requires. Subjectively, it doesn't feel like there's enough value to implement it right now.

Reply:
> I picked it because I was already thinking of using the position of the first recognized XML tag, and it was more straight forward to implement if lower score is better.

OK, that's fine.  It's not clearly better if higher or lower is better, so I'm fine with what we currently have.



> > 
> 
> There might be some value on a development machine where a robot description is ambiguous, or an installed plugin takes a really long time to do `might_handle()`. On a robot itself there's less value because presumably someone has chosen to install only the plugins the robot requires. Subjectively, it doesn't feel like there's enough value to implement it right now.

OK, I just thought I'd bring up the point.  I'm fine with leaving that for later if we need it.